pkg/adaptor/errors.go

```diff
@@ -0,0 +1,27 @@
+//
+// Copyright (c) 2019-2020 Red Hat, Inc.
+// This program and the accompanying materials are made
+// available under the terms of the Eclipse Public License 2.0
+// which is available at https://www.eclipse.org/legal/epl-2.0/
+//
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+//
+
+package adaptor
+
+// DownloadMetasError represents an error that occurs while downloading plugin meta.yamls
+// This error wraps the underlying error that caused the failure.
+type DownloadMetasError struct {
+	Plugin string
+	Err    error
+}
+
+var _ error = (*DownloadMetasError)(nil)
+
+func (e *DownloadMetasError) Error() string {
+	return "Failed to download plugin meta.yaml for " + e.Plugin
```

Comment:
Operator admin could check logs for more details:
```
2020-12-03T13:12:38.626+0200    INFO    controllers.Component   Failed to download plugin metas {"Request.Namespace": "devworkspace-controller", "Request.Name": "components-workspace827d4c63d6bb4955-plugins", "err": "failed to fetch plugin meta.yaml from URL 'http://che-plugin-registry-devworkspace-controller.apps.ci-ln-27h2cmk-f76d1.origin-ci-int-gce.dev.openshift.com/v3/plugins/eclipse/che-does-not-exist/latest/meta.yaml': Downloading http://che-plugin-registry-devworkspace-controller.apps.ci-ln-27h2cmk-f76d1.origin-ci-int-gce.dev.openshift.com/v3/plugins/eclipse/che-does-not-exist/latest/meta.yaml failed. Status code 404. Response body: <!DOCTYPE HTML PUBLIC \"-//IETF//DTD HTML 2.0//EN\">\n<html><head>\n<title>404 Not Found</title>\n</head><body>\n<h1>Not Found</h1>\n<p>The requested URL was not found on this server.</p>\n</body></html>\n"}
```
But from a user-perspective, it may be really difficult to figure out the cause, like if default plugin registry is misconfigured, operator has bug and construct wrong URL. So
maybe it makes sense to add here at least plugin URL, like `Failed to download plugin registry meta.yaml for e.Plugin. Please check e.URL`

Reply:
This is actually tricky -- plugin resolution is handled completely by the plugin broker, which currently does not offer a convenient way of finding the URL we'll eventually use to try and download the plugin. We could potentially grab it from the underlying error message, but it's tricky as that tends to log the full response, which in the case of OpenShift is an HTML page sometimes.

The correct way to surface the URL we're trying would be to break [this section](https://github.com/eclipse/che-plugin-broker/blob/92c3b00094475b3a157de6dd892ec28119de763f/utils/metadata_resolver.go#L36-L46) of the broker out into a separate function, release the broker, and then use that here.